fl4health/server/tabular_feature_alignment_server.py

```diff
@@ -0,0 +1,143 @@
+
+    def fit(self, num_rounds: int, timeout: Optional[float]) -> History:
+        """Run federated averaging for a number of rounds."""
+
+        assert isinstance(self.strategy, BasicFedAvg)
+
+        # Before the normal fitting round begins, the server provides all clients
+        # the feature information needed to perform feature alignment. Then the server
+        # gathers information from the clients that is necessary for initializing the global model.
+        if not self.initial_polls_complete:
+
+            # If the server does not have the needed feature info a priori,
+            # then it requests such information from the clients before the
+            # very first fitting round.
+            if self.tab_features_info is None:
+                # A random client's feature information is selected as the source of truth for feature alignment.
+                feature_info = self.poll_clients_for_feature_info(timeout)
+
+                rand_idx = random.randint(0, len(feature_info) - 1)
+
+                feature_info_source = feature_info[rand_idx]
+            # If the server already has the feature info, then it simply sends it to the clients.
+            else:
+                log(
+                    INFO,
+                    "Features information source already specified. Sending to clients to perform feature alignment.",
+                )
+                feature_info_source = self.tab_features_info.to_json()
+
+            # the feature information is sent to clients through the config parameter.
+            self.config[FEATURE_INFO] = feature_info_source
+            self.format_info_gathered = True
+
+            self.strategy.on_fit_config_fn = partial(fit_config, self.config, self.format_info_gathered)
+
+            # Now the server waits until feature alignment is performed on the clients' side
+            # and subsequently requests the input and output dimensions, which are needed for initializing
+            # the global model.
+            input_dimension, output_dimension = self.poll_clients_for_dimension_info(timeout)
+            log(INFO, f"input dimension: {input_dimension}, output dimension: {output_dimension}")
+            self.strategy.initial_parameters = self.initialize_parameters(input_dimension, output_dimension)
+            self.initial_polls_complete = True
+
+        # Normal federated learning rounds commence after all clients' features
+        # are aligned and global model is initialized.
+        return super().fit(num_rounds=num_rounds, timeout=timeout)
+
+    def poll_clients_for_feature_info(self, timeout: Optional[float]) -> List[str]:
+        log(INFO, "Feature information source unspecified. Polling clients for feature information.")
+        assert isinstance(self.strategy, BasicFedAvg)
+        client_instructions = self.strategy.configure_poll(server_round=1, client_manager=self._client_manager)
```

Comment:
This is a bit of a hack, but configure_poll selects all available clients. We actually don't want all clients, we want to randomly select a single on to get feature info from. So I would suggest that rather than having all clients send their feature information and we just select one, we randomly select one from the `client_instructions` using `random.choice`. This will then poll that randomly selected client for features, results can be asserted to be of length 1 and feature_info will just be a string rather than a list.

Reply:
Yes I think this a good idea! Originally I polled all of the clients thinking we might want to select multiple ones and merge their information, but even in that case we can just use `random.choice` on `client_instructions`.

Reply:
I think this also applies to the `poll_clients_for_dimension_info` method, so I'll make these changes there as well.